### DISCUSSION TITLE
Best practice for connect and close of async client

### DISCUSSION THREAD

**Discussion by heinehenning** [pymodbus-dev/pymodbus#1756] [Help] [answered]
Hi all, 

I am currently building a modbus client with the AsyncModbusTcpClient.

The examples work very well. But I wonder, is it best practice to open and close the connection on every polling operation, like shown in the examples:

```
async def run_async_client(client, modbus_calls=None):
    """Run sync client."""
    _logger.info("### Client starting")
    await client.connect()
    assert client.connected
    if modbus_calls:
        await modbus_calls(client)
    client.close()
    _logger.info("### End of Program")
```

I thought it would be better to open the connection once and close the connection at a later point whenever communcation should stop.

Would be great to get a hint how that could be implemented with the async client.

Thanks in advance!

**Comment by janiversen** [comment, marked as answer]
of course not, see e.g. client_calls.py.

You connect in the beginning and close when finished.


Actually the example says that, it's modbus_calls not modbus_call

I am not what sure hint you want the code is right there in the example,
1: create client
2: connect
3: run all the calls you want
4: close

**Comment by heinehenning** [comment]
Thanks @janiversen for your answer.
I think my problem was with asyncio. I tried to await for the client.close() method. And that always produced exceptions.
Closing the client without await works.

**Reply by janiversen** [reply]
Just read the documentation, that helps.

The documentation, as well as our examples, how to use and not to use close().
